_drafts/device-logging.md

```diff
@@ -0,0 +1,295 @@
+
+A major impedance to reading logs is when the exact same string is re-used multiple times.
+
+```markdown
+V bort    : Payment started: payment_session_id=0xFF
+```
+
+```kotlin
+fun startPayment() {
+    Logger.v("Payment started: payment_session_id=$id")
+    // ...
+}
+
+fun retryPayment() {
+    Logger.v("Payment started: payment_session_id=$id")
+    // ...
+}
+```
+
+If you see this log while you're trying to diagnose an issue, you will probably have no clue which line of code it’s actually referring to. So it’s important to spend the extra time to make sure that all our logs are unique from one another.
+
+### Log for Computers First
+
+We'll never really stop reading logs, but it can become a chore to read 1000s of lines of logs, or look to see if multiple sets of logs from multiple devices may be experiencing the same issue. One of the best ways to get out of the business of reading logs, is to have our computers read it for us, and display it in a more digestible format.
+
+At my past companies, we developed shared command line tools that would read log files and print out a text summary of all the exceptions, any network connectivity issues detected, and various other useful information. The benefit of this was massive; not only did the tool analyze logs much faster than we could individually, but having a shared solution meant everyone benefited from the time savings, we all agreed upon the interpretation of different log messages, and we could add new features to the tool as the team grew and new scenarios were discovered.
+
+To facilitate that more easily, it would’ve helped if we could rely on the format of logs to be much more machine readable.
+
+#### Avoid Multiline Logs
+
+Multiline logs are a particularly painful problem for computers to parse. Most system expect logs to all be on a single line, and tracking logs across multiple lines requires retaining state specific to that error message just to parse that error.
+
+```markdown
+E bort    : Network error: -55
+E bort    :     RSSI: 61
+E bort    :     SSID: "Home-Guest"
+```
+
+On one line, it’s slightly less visually readable, but vastly more easily machine parsable.
+
+```markdown
+E bort    : Network error: -55, RSSI: 61, SSID: "Home-Guest"
+```
+
+#### Structured Logs
+
+We can take another page out of server logging best practices [^structuredlogging], and use structured logging. Logs by default are unstructured strings. If we logged more often in a structured format that was easily machine parsable, we could greatly simplify parsing and organizing the data. This is especially useful when we want to eventually parse specific information in the log, otherwise we'd have to write a Regex to parse the log.
+
+```markdown
+E bort    : {"Network error": "-55", "RSSI": 61, "SSID": "Home-Guest"}
+```
+
+JSON is the most common format for structured logs, and can be written manually in code fairly quickly. It's also fairly readable compared to the unstructured log, too.
```

Comment:
Maybe worth calling out that many platforms have a maximum line length for logs which is easy to blow past in json.

AOSP has some best practices in their [coding guidelines](https://source.android.com/docs/setup/contribute/code-style#log-sparingly) - one of which is ironically "Log Sparingly" and maybe something worth pulling into this article. They recommend log message lenghts up to 80-100 chars and say to avoid lengths of 130-160 **including the tag**.

Reply:
Good idea about the line length.

Mmm I actually disagree with a lot of general guidance around logging, especially the log sparingly part. It's a totally nonsensical because there's already 200 services logging all the time in AOSP, so they're not even following their own guidance. And I particularly dislike the maximum line length guidance. It's a meaningless recommendation and AOSP exceeds the max all the time as well. The idea should just be, be as concise as you can without losing the information, and log multiple times you if a single line isn't sufficnet. All of this is probably trauma from Py trying to "fix logging".

I actually really like most of the points in the NOTES part, but then it ends with "The golden rule of logging is that your logs may not unnecessarily push other logs out of the buffer, just as others may not push out yours." which is just ??? - the kind of fluff I dislike. 

Honestly you could write a whole separate article on every single idea around improving logging.